botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1027 @@
+                {
+                "type": "tree",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "rules": [
+                    // Abbreviated for clarity
+                ]
+                },
+                {
+                "type": "endpoint",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "endpoint": {
+                    "url": "{parameterName}.amazonaws.com"
+                }
+                },
+                {
+                "type": "error",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "error": "an error message"
+                }
+            ]
+            }
+        ]
+    }
+    """
+
+    def __init__(self, rules, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.rules = [
+            RuleCreator.create(
+                assignments=self.assignments, provider=self.provider, **rule
+            )
+            for rule in rules
+        ]
+
+    def evaluate(self):
+        """If a tree rule's conditions evaluate successfully, iterate over its
+        subordinate rules and return a result if there is one. If any of the
+        subsequent rules are trees, the function will recurse until it reaches
+        an error or an endpoint rule.
+
+        :returns: An endpoint or error object.
+        """
+        if self.evaluate_conditions() is True:
+            for rule in self.rules:
+                evaluation = rule.evaluate()
+                if evaluation:
+                    return evaluation
+        return None
+
+
+class RuleCreator(Enum):
+    """A helper class that creates rule objects."""
+
+    endpoint = EndpointRule
+    error = ErrorRule
+    tree = TreeRule
+
+    @classmethod
+    def create(cls, **kwargs):
+        """Given a rule dictionary, locate the `type` attribute, using it to
+        determine the type of rule object to create. Then create and return it.
+
+        :type kwargs: dict
+        :params kwargs: An arbitrary dictionary containing required parameters
+        for different types of rule objects i.e. conditions, rules, errors,
+        endpoints etc.
+
+        :returns: A TreeRule/EndpointRule/ErrorRule object
+        """
+        rule = getattr(cls, kwargs.pop('type'))
+        return rule.value(**kwargs)
+
+
+class ParameterDefinition:
+    """A class that houses the spec of an individual parameter defined in a
+    rule set object i.e.
+    {
+        "Region": {
+            "builtIn": "AWS::Region",
+            "required": true,
+            "type": "String"
+        }
+    }
+    """
+
+    def __init__(
+        self,
+        name,
+        type,
+        documentation=None,
+        builtIn=None,
+        default=None,
+        required=None,
+        deprecated=None,
+    ):
+
+        self.name = name
+        self.type = type
```

Comment:
`type` is a builtin function, we may want to choose something different for the name here. `parameter_type`?

Reply:
yeah i gave this a lot of thought. i just really liked the idea of being imperative with `kwargs` since the the key name in a rule set is `type`, but i can change it